src/avro/decode.rs

```diff
@@ -207,23 +224,63 @@ pub fn decode<'a, R: AsyncRead + Unpin + Send>(
 
                 Ok(Value::Map(items))
             }
-            Schema::Union(ref inner) => {
-                let index = zag_i64(reader).await?;
+            SchemaPiece::Union(inner) => {
+                let index = zag_i64(reader).await? as usize;
                 let variants = inner.variants();
-                match variants.get(index as usize) {
-                    Some(variant) => decode(variant, reader)
+                match variants.get(index) {
+                    Some(variant) => decode(schema.step(variant), reader)
                         .await
-                        .map(|x| Value::Union(Box::new(x))),
+                        .map(|x| Value::Union(index, Box::new(x))),
                     None => Err(DecodeError::new("Union index out of bounds").into()),
                 }
             }
-            Schema::Record { ref fields, .. } => {
+            SchemaPiece::ResolveIntLong => zag_i32(reader).await.map(|x| Value::Long(x as i64)),
+            SchemaPiece::ResolveIntFloat => zag_i32(reader).await.map(|x| Value::Float(x as f32)),
+            SchemaPiece::ResolveIntDouble => zag_i32(reader).await.map(|x| Value::Double(x as f64)),
+            SchemaPiece::ResolveLongFloat => zag_i64(reader).await.map(|x| Value::Float(x as f32)),
```

Comment:
Yep, agreed

Reply:
Actually the int/long |-> float ones can't (I guess f32 can't represent every integer).

The avro spec doesn't say what to do in case of overflow, so I guess silently truncating is the best you can hope for.